src/mqtt_client.rs

```diff
@@ -216,53 +186,63 @@ where
         topic: &'a str,
         properties: &[Property<'b>],
     ) -> Result<(), Error<TcpStack::Error>> {
-        if self.connection_state.state() != &States::Active {
+        if self.sm.state() != &States::Active {
             return Err(Error::NotReady);
         }
 
+        let ClientContext {
+            network,
+            session_state,
+            ..
+        } = self.sm.context_mut();
+
         // We can't subscribe if there's a pending write in the network.
-        if self.network.has_pending_write() {
+        if network.has_pending_write() {
             return Err(Error::NotReady);
         }
 
-        let packet_id = self.session_state.get_packet_identifier();
+        let packet_id = session_state.get_packet_identifier();
 
         let mut buffer: [u8; MSG_SIZE] = [0; MSG_SIZE];
         let packet = serialize::subscribe_message(&mut buffer, topic, packet_id, properties)?;
 
-        self.network.write(packet).and_then(|_| {
+        network.write(packet).and_then(|_| {
             info!("Subscribing to `{}`: {}", topic, packet_id);
-            self.session_state
+            session_state
                 .pending_subscriptions
                 .push(packet_id)
                 .map_err(|_| Error::Unsupported)?;
-            self.session_state.increment_packet_identifier();
+            session_state.increment_packet_identifier();
             Ok(())
```

Comment:
I'd push this code down into `ClientContext`.

Reply:
I don't think we can push this all the way down without introducing a whole ton of boilerplate code.

Specifically, we need:
1. To verify that `subscribe()` occurs after we are connected, since we can't subscribe before we're connected via MQTT.
2. Provide the `subscribe()` mechanism to the `MqttClient` object, so that it can be called in both `minimq.client().poll()` and in `minimq.client()`.

If we pushed this code down into `ClientContext`, we'd have to expose a new `subscribe` function on the `MqttClient` that checked the state and then just called into the `ClientContext::subscribe()` function with all of the duplicated arguments.

We can do that, it just felt like a lot of copy-paste boilerplate to me.

Reply:
Yes, there would be a `ClientContext::subscribe()` and you would call that from a two-line function `MqttClient::subscribe()`. But that's instead of the all the destructuring of `context_mut()` in `MqttClient`.
My goal is to contain `network` and `session_state` code within the context that owns them and not leak their management code trough `pub(crate) network` and destructuring.

Reply:
Yeah, I'm in agreement and in the process of cleaning some of it up.

Reply:
For cases where we don't need `state()` or `process_event()`, maybe a reasonable compromise is to just make it `MqttClient::context_mut().subscribe()`.